File: code/cpp/46.cpp
//https://leetcode.com/problems/permutations/
class Solution {
public:
    void foo(vector<int> nums, vector<vector<int>>& res, int index) {
        if (index >= nums.size()) {
            res.push_back(nums);
            return;
        }
        for (int i = index; i < nums.size(); i++) {
            swap(nums[index], nums[i]);
            foo(nums, res, index + 1);
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        foo(nums, res, 0);
        return res;
    }
};

File: code/python/46.py
#https://leetcode.com/problems/permutations/
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        def foo(index):
            if index == len(nums):
                res.append(nums[:])
                return
            for i in range(index, len(nums)):
                nums[index], nums[i] = nums[i], nums[index]
                foo(index + 1)
                nums[index], nums[i] = nums[i], nums[index]

        res = []
        foo(0)
        return res
